### ISSUE TITLE
sally-english

### ISSUE THREAD

**Issue by brianteeman** [ryersondmp/sa11y#9]
Normally I would do a pull request but I am assuming that you are developing outside of github so that wouldnt be much help.

In the text below "a image" should be "an image" and "photo" is present twice.

https://github.com/ryersondmp/sa11y/blob/a486372133353c1de4e9ab5fb768bf7d843953e6/src/sa11y-english.js#L65-L85

**Comment by adamchaboryk** [comment]
Hi Brian,

Yes, I am still actively fixing little bugs and such. 😅

Thanks for the catch with "an image". Many of them are repeated on purpose. If the alt tag _starts_ with any of these words, it will flag it as a warning. 

* "image of",
*  "an image of",
*  "graphic of",
*  "a graphic of",
*  "picture of",
*  "a picture of",
*  "photo of",
*  "a photo of",
*  "pic of",
*  "a pic of"

**Fixed:**

//Alt Text stop words.
`const sa11ySuspiciousAltWords = [
    "image of",
    "an image of",
    "graphic of",
    "a graphic of",
    "picture of",
    "a picture of",
    "photo of",
    "a photo of",
    "pic of",
    "a pic of"
];`

**Comment by brianteeman** [comment]
Instead of checking if it _starts with_ why not check if it _includes_ ??

That way you you could reduce the array to just 
`const sa11ySuspiciousAltWords = [ "image", "graphic", "picture", "photo", "pic" ];
`
You would also catch "this is a picture of .." etc

**Comment by adamchaboryk** [comment]
I decided to go with _starts with_ to minimize the number of warnings a content author might see. In certain contexts those words might be helpful/necessary. Although usually only unnecessary/verbose when these filler words are the start of the alt text.

E.g.

* **Old photos:** _Black and white photo of Albert Einstein sticking out his tongue._  **or** _A grainy photo of a younger President John Doe visiting Buckingham Palace in 1965._
* **Images of text:** _Quote: To be or not to be that is the question. Photo of a man holding a skull._

Etc. 

Thoughts?

**Comment by brianteeman** [comment]
I take your point but I guess I am looking at this from a different angle - that  of translators.
Also this is only a "warning" or "advice" and not an error

**Comment by adamchaboryk** [comment]
Still thinking about this more. I ultimately want to minimize the number of warnings a content author can see.

I think I have an over engineered solution:

Where `sa11ySuspiciousAltWordsPosition` is a global configuration variable with two potential values:

1. `start` = The start of an alt attribute starts with "image of", "graphic of", etc.
2. `any` = Location of potentially redundant word doesn't matter.

**sa11y-english.js**
`````javascript
const sa11ySuspiciousAltWordsPosition = "any"; // "start" or "any"

const sa11ySuspiciousAltWordsArray = {
    start: ["image of", "an image of", "graphic of", "a graphic of", "picture of", "a picture of", "photo of", "a photo of", "pic of", "a pic of"],
    any: ["image", "graphic", "picture", "photo", "pic"],
}
`````

**sa11y.js**
`````javascript
if (sa11ySuspiciousAltWordsPosition === "start") {
        $.each(sa11ySuspiciousAltWordsArray[sa11ySuspiciousAltWordsPosition], function (index, word) {
                 if (alt.toLowerCase().indexOf(word) == 0) {
                       hit[1] = word;
                 }
        });
} else if (sa11ySuspiciousAltWordsPosition === "any") {
         $.each(sa11ySuspiciousAltWordsArray[sa11ySuspiciousAltWordsPosition], function (index, word) {
                if (alt.toLowerCase().indexOf(word) >= 0) {
                         hit[1] = word;
                   }
         });
}
`````

Overkill? Am I over thinking this too much? I need sleep. Lol.

**Comment by brianteeman** [comment]
> Overkill? Am I over thinking this too much?

yes I think it is.

**Comment by itmaybejj** [comment]
I think I'm with brianteeman. I can't tell you what the frequency is, but I suspect "J smith photo" and "photo of J Smith" both appear at times. 

My long term plan to address this is to add some sort of "dismiss" or "this one is OK" button for these sorts of fuzzier rules, that drops the image SRC and/or alt text as an identifier into a localStorage array -- probably an array like ["page_url","rule_id","rule_specific_element_identifier"]. But hecka lotta coding, since each rule that I want to be suppressible will need to generate its identifier and then check to see if there's a match in the array before pushing to the alert list. And even then...if I only use localStorage, old alerts will return someday...

**Comment by adamchaboryk** [comment]
Thank you both for talking sense into me! I have reduced the array to `["image", "graphic", "picture", "photo"]` - as you said, it is a warning after all. 

@itmaybejj, that sounds pretty ambitious. 😅 I tried my best to minimize the number of warnings for those fuzzier rules. E.g. if link text starts with https:// - I only flag it as an issue after 40 or so characters. I personally haven't received any feedback or complaints about excessive warnings (...yet...).

>Feel free to do that work so I can lift it 🤣

_Yeah, that’s gonna be a no from me dawg._

Revised tooltip verbiage:
<img width="660" alt="Screenshot of two tooltips with redundant alt text." src="https://user-images.githubusercontent.com/26335725/125152205-240a0300-e119-11eb-9876-9fc31b24ef24.png">

Will be included in next release!
